examples/cosmwasm/zero-to-hero/vote.qnt

```diff
@@ -0,0 +1,284 @@
+
+  type MessageInfo = {
+    sender: str
+  }
+
+  type InstantiateMsg = {
+    admin: str
+  }
+
+  // 08 - ExecuteMsg
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/08%20-%20ExecuteMsg
+
+  type CreatePoll = {
+    poll_id: UUID,
+    question: str,
+    options: List[str]
+  }
+
+  type Vote = {
+    poll_id: UUID,
+    vote: str,
+  }
+
+  // 09 - Execute 1
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/09%20-%20Execute%201
+
+  type Result = {
+    error: str,
+    cs: ContractState
+  }
+
+  // 12 - QueryMsg
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/12%20-%20QueryMsg
+
+  type QueryAllPolls = {}
+  type QueryPoll = { poll_id: str }
+  type QueryVote = { poll_id: str, address: Addr }
+
+  /*** Functional layer ***/
+
+  // 06 - Instantiate
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/06%20-%20Instantiate
+
+  def instantiate(cs: ContractState, info: MessageInfo,
+                  msg: InstantiateMsg): ContractState = {
+    val admin = if (msg.admin != None) msg.admin else info.sender
+    // TODO: deps.api.addr_validate(&admin)?
+    { ...cs, config: { admin: admin } }
+  }
+
+  // 09 - Execute 1
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/09%20-%20Execute%201
+
+  def execute_create_poll(cs: ContractState, info: MessageInfo, poll_id: UUID,
+                          question: str, options: List[str]): Result = {
+    if (options.length() > 10) {
+      { cs: cs, error: "TooManyOptions" }
+    } else {
+      val opts = options.foldl([], (res, opt) => res.append((opt, 0)))
+
+      val poll = {
+        creator: info.sender,
+        question: question,
+        options: opts
+      }
+      
+      // CosmWasm: POLLS.save(deps.storage, poll_id, &poll)?;
+      { cs: { ...cs, polls: cs.polls.put(poll_id, poll) }, error: "" }
+    }
+  }
+
+  // 10 - Execute 2
+  // https://github.com/Callum-A/cosmwasm-zero-to-hero/tree/main/10%20-%20Execute%202
+
+  def execute_vote(cs: ContractState, info: MessageInfo, poll_id: UUID,
+                   vote: str): Result = {
+    // CosmWasm: POLLS.may_load(deps.storage, poll_id.clone())?;
+    if (not(cs.polls.keys().contains(poll_id))) {
+      // no such poll
+      { cs: cs, error: "Unauthorized" }
+    } else {
+      val poll = cs.polls.get(poll_id)
+      val _key = (info.sender, poll_id)
+
+      val updated = if (not(cs.ballots.keys().contains(_key))) {
+        // sender hasn't voted on this poll; simply add the ballot
+        (poll.options, cs.ballots.put(_key, { option: vote }))
+      } else {
+        // sender already voted on the poll; adjust vote count for removing their old vote
+        val ballot = cs.ballots.get(_key)
+        val optionsDecr = poll.options.foldl([], (res, opt) => {
+          // FIXME: in the original, this only adjusts the _first_ option in the list
+          if (opt._1 == ballot.option) res.append((opt._1, opt._2 - 1)) else res.append(opt)
```

Comment:
How does it work in the original contract? It looks to me like an inplace modification, but should not the borrow checker handle that somehow?